torch_xla/csrc/runtime/ifrt_computation_client.h

```diff
@@ -0,0 +1,249 @@
+                                 std::string device, xla::Shape shape,
+                                 xla::OpSharding sharding) override;
+
+  DataPtr CopyToDevice(DataPtr data, std::string dst) override;
+
+  std::vector<ComputationPtr> Compile(
+      std::vector<CompileInstance> instances) override;
+
+  std::vector<DataPtr> ExecuteComputation(
+      const Computation& computation, absl::Span<const DataPtr> arguments,
+      const std::string& device,
+      const ExecuteComputationOptions& options) override;
+
+  std::vector<std::vector<DataPtr>> ExecuteReplicated(
+      const Computation& computation,
+      const std::vector<std::vector<DataPtr>>& arguments,
+      absl::Span<const std::string> devices,
+      const ExecuteReplicatedOptions& options) override;
+
+  size_t GetNumDevices() const override;
+
+  std::string GetDefaultDevice() const override;
+
+  std::vector<std::string> GetLocalDevices() const override;
+
+  std::vector<std::string> GetAllDevices() const override;
+
+  int GetProcessIndex() const override { return client_->process_index(); };
+
+  int GetNumProcesses() const override;
+
+  const absl::flat_hash_map<
+      std::string, torch_xla::runtime::ComputationClient::DeviceAttribute>&
+  GetDeviceAttributes(const std::string& device) override;
+
+  void SetReplicationDevices(
+      std::shared_ptr<std::vector<std::string>> devices) override;
+
+  std::shared_ptr<std::vector<std::string>> GetReplicationDevices() override;
+
+  void PrepareToExit() override { return; };
+
+  void WaitDeviceOps(const std::vector<std::string>& devices) override;
+
+  std::map<std::string, Metric> GetMetrics() const override;
+
+  // NOT IMPLEMENTED
+
+  MemoryInfo GetMemoryInfo(const std::string& device) override {
+    XLA_ERROR() << __FUNCTION__ << " not implemented";
+  };
+
+ private:
+  std::shared_ptr<xla::ifrt::PjRtClient> client_;
+  // global_ordinals_ tracks a map from PjRtDeviceId to the device's
+  // dense global ordinal.
+  std::unordered_map<int, int> global_ordinals_;
+  std::unordered_map<std::string, xla::PjRtDevice* const> string_to_device_;
+  std::shared_ptr<std::vector<std::string>> replication_devices_;
+  std::unordered_map<std::string, std::unique_ptr<std::shared_mutex>>
+      device_locks_;
+
+  xla::PjRtDevice* StringToPjRtDevice(const std::string& device);
+  std::shared_lock<std::shared_mutex> lock_device_shared(
+      const std::string& device);
+  std::unique_lock<std::shared_mutex> lock_device(const std::string& device);
+
+  std::string PjRtDeviceToString(xla::PjRtDevice* const device) const;
+  std::vector<std::string> PjRtDevicesToString(
+      absl::Span<xla::PjRtDevice* const> devices) const;
+
+  struct IfrtData : public Data {
+    IfrtData(std::string device, xla::Shape device_shape)
+        : Data(std::move(device), std::move(device_shape)) {}
+
+    IfrtData(std::string device, xla::Shape device_shape,
+             tsl::RCReference<xla::ifrt::Array> buffer,
```

Comment:
Some pointer to verify -- We are wrapping `xla::ifrt::pjrt_client`, and that probably means that you want to work with the same `PjRtBuffer` instead of the `ifrt::Array` [link](http://google3/third_party/tensorflow/compiler/xla/python/pjrt_ifrt/pjrt_client.h;l=43;rcl=558837006). 

I would think we should re-use `PjRtData` and `PjRtShardedData` as-is, and later add a new data type `IfRtArray`, when it's supported in `ifrt::pjrt_client` ?

Reply:
Actually, I see array is already being supported in the client there.